### PR TITLE
Deprecate whereNotNull from IterableNullableExtensions

### LINKED ISSUE #273: rewrite whereNotNull without sync*
Current implementation of `sync*` is suboptimal and we might not get to fixing its performance in Dart 3.0 so we should probably avoid using `sync*` for now.

See also https://github.com/dart-lang/sdk/issues/51806

### PR DESCRIPTION
**Pull request by oprypin**
Dart SDK since 3.0 has an exact equivalent extension `nonNulls` in the core.

---

* References #330
* Closes #273
* Closes #274

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by natebosch** [comment]
cc @bwilkerson - would it be feasible to add support in dart_fix.yaml to replace a method with no arguments with a field? I don't think we currently have a way to automate the fix we need:

```dart
import 'package:collection/collection.dart';

void main() {
  [].whereNotNull().join('');
}
```

Should become

```dart
void main() {
  [].nonNulls.join('');
}
```

The transform I think I want, but is currently an error:

```yaml
  - title: 'Replace whereNotNull() with core nonNulls'
    date: 2024-01-25
    element:
      uris: ['collection.dart']
      method: 'whereNotNull'
      inExtension: 'IterableExtension'
    changes:
      - kind: 'replacedBy'
        newElement:
          uris: ['dart:core']
          field: 'nonNulls'
          inClass: 'NullableIterableExtensions'
```

File: test/extensions_test.dart [hunk 1]
```diff
@@ -866,23 +866,6 @@ void main() {
         });
       });
     });
-    group('of nullable', () {
```
Comment (kevmoo) [inline test/extensions_test.dart:869]: We can leave the tests while we still ship it. Just ignore the lint!

Comment (oprypin) [reply test/extensions_test.dart:869]: CI fails, not possible to ignore presumably

Comment (kevmoo) [reply test/extensions_test.dart:869]: You can ignore in the source code!

Comment (kevmoo) [reply test/extensions_test.dart:869]: bingo!

**Comment by kevmoo** [comment]
Are we going to move forward with this – or should we wait until after 1.19.0? @natebosch

**Comment by lrhn** [comment]
We can remove it in the next major version, which may be the one which switches to Dart 3.0. Maybe.

**Comment by kevmoo** [comment]
Well the constraint @ master right now is `^3.1.0` – so this would be as good a time as any.

**Review by kevmoo — APPROVED** [review]

**Pull request merged**